tests/dragonfly/generator.py

```diff
@@ -0,0 +1,245 @@
+
+        cmds = []
+        while len(cmds) < self.batch_size:
+            if len(actions) == 0:
+                actions = random.choices(
+                    list(Action), weights=self.action_weigts(), k=50)
+
+            cmd, delta = self.make(actions.pop())
+            if cmd is not None:
+                cmds.append(cmd)
+                self.key_cnt += delta
+
+        return cmds, self.key_cnt/self.key_cnt_target
+
+
+class TestingExecutor:
+    def __init__(self, client, target_bytes):
+        self.client = client
+
+        # Count number of expected keys. Account for memory overhead.
+        max_multikey = 5
+        batch_size = 4000
+        val_size = 500
+        target_keys = target_bytes // val_size
+        print("target keys", target_keys)
+        self.gen = TestingGenerator(
+            val_size, batch_size, target_keys, max_multikey)
+
+        self.task = None
+        self.task_target_rel = None
+        self.batches = deque()
+        self.max_batches = 3
+
+    async def run(self, target_times=None, target_devitation=None):
+        cpu_time = 0
+        submitted = 0
+        while True:
+            # While we're waiting lets pre-generate data
+            while self.task is not None and not self.task.done() and len(self.batches) < self.max_batches:
+                s = time.time()
+                self.batches.append(tuple(self.gen.generate()))
+                cpu_time += (time.time()-s)
+
+            # Await task
+            if self.task is not None:
+                await self.task
+                submitted += 1
+                #print("Submitted", self.task_target_rel)
+                if target_times is not None and submitted >= target_times:
+                    return
+                if target_devitation is not None and abs(1-self.task_target_rel) < target_devitation:
+                    print("done", submitted, "time", cpu_time)
+                    return
+
+            # Generate or fetch new data
+            if len(self.batches) > 0:
+                cmds, target_rel = self.batches.popleft()
+            else:
+                print("generating on demand :(")
+                cmds, target_rel = self.gen.generate()
+
+            # Schedule task
+            pipe = self.client.pipeline(transaction=False)
+            for cmd in cmds:
+                pipe.execute_command(cmd)
+
+            self.task = asyncio.create_task(pipe.execute())
+            self.task_target_rel = target_rel
+
+    async def checkhash(self, client=None):
+        if client is None:
+            client = self.client
+
+        all_keys = sorted(list(self.gen.keys()))
+        hashes = []
+
+        for chunk in chunked(self.gen.batch_size, all_keys):
+            pipe = client.pipeline(transaction=False)
+            for k in chunk:
+                pipe.execute_command(f"DUMP {k}")
+            res = await pipe.execute()
+            hashes.append(hash(tuple(res)))
+
+        return hash(tuple(hashes))
```

Comment:
We need a way to quickly compare two instances states (to make sure the replica and master have the same data).

So my suggestion is iterating all the values in pre-defined order (sorted by keys), represent each as a string and calculate the cumulative hash.

We should do this fully on server and return only the hash, so we need a new debug command for this. Are there maybe any easier options?